IonDotnet/Internals/Tree/UserTreeReader.cs

```diff
@@ -0,0 +1,137 @@
+            if (!NextHelperUser())
+            {
+                _current = null;
+                return IonType.Null;
+            }
+            _current = _next;
+            _next = null;
+            return _current.Type();
+        }
+
+        bool NextHelperUser()
+        {
+            if (_eof) return false;
+            if (_next != null) return true;
+
+            ClearSystemValueStack();
+
+            // read values from the system
+            // reader and if they are system values
+            // process them.  Return when we've
+            // read all the immediate system values
+            IonType next_type;
+            while (true)
+            {
+                next_type = NextHelperSystem();
+
+                if (_top == 0 && _parent.Type() == IonType.Datagram)
+                {
+                    if (IonType.Symbol == next_type)
+                    {
+                        var sym = _next;
+                        if (sym.IsNull)
+                        {
+                            // there are no null values we will consume here
+                            break;
+                        }
+                        int sid = sym.SymbolValue.Sid;
+                        if (sid == -1) // if sid is unknown
+                        {
+                            String name = sym.SymbolValue.Text;
+                            if (name != null)
+                            {
+                                sid = _systemSymbols.FindSymbolId(name);
+                            }
+                        }
+                        if (sid == ION_1_0_SID && _next.GetTypeAnnotations().Count == 0)
+                        {
+                            // $ion_1_0 is read as an IVM only if it is not annotated
+                            ISymbolTable symbols = _systemSymbols;
+                            _currentSymtab = symbols;
+                            PushSymbolTable(symbols);
+                            _next = null;
+                            continue;
+                        }
+                    }
+                    else if (IonType.Struct == next_type && _next.HasAnnotation("$ion_symbol_table"))
+                    {
+                        // read a local symbol table
+                        IIonReader reader = new UserTreeReader(_next, _catalog);
+                        ISymbolTable symtab = ReaderLocalTable.ImportReaderTable(this, _catalog, true);
+                        _currentSymtab = symtab;
+                        PushSymbolTable(symtab);
+                        _next = null;
+                        continue;
+                    }
+                }
+                // if we get here we didn't process a system
+                // value, if we had we would have 'continue'd
+                // so this is a value the user gets
+                break;
+            }
+            return (next_type != IonType.Null);
+        }
+
+        private int _symbol_table_top = 0;
```

Comment:
Should we group these variables to the top?

Reply:
Done.